rfd/0197-better-async-state-management-with-tanstack-query.md

```diff
@@ -0,0 +1,266 @@
+
+  accessManagementService
+    .fetchAccessList(accessListId)
+    .then(fetchedAccessList => {
+      const [modifiedAccessList, newPerms] = modifyAccessList(
+        fetchedAccessList,
+        accessLists,
+        ctx
+      );
+      setAccessList(modifiedAccessList);
+      setPerms(newPerms);
+      scopedAttempt.setAttempt({ status: 'success' });
+    })
+    .catch((e: Error) =>
+      scopedAttempt.setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+This then has to be called with an effect that watches the access list ID:
+
+```tsx
+// The accessListId can change if a user clicks on a different
+// access list in the notification dropdown.
+useEffect(() => {
+  if (!accessList || accessList.id !== accessListId) {
+    fetchAccessList();
+  }
+}, [accessListId]);
+```
+
+#### Example mutation
+
+```tsx
+const { attempt, setAttempt } = useAttempt();
+const isDisabled = attempt.status === 'processing';
+
+function onOk() {
+  setAttempt({ status: 'processing' });
+  accessManagementService
+    .deleteAccessList(accessListId)
+    .then(() =>
+      history.push(cfg.getAccessListManagementRoute(), {
+        deletedAccessListId: accessListId,
+      })
+    )
+    .catch((e: Error) =>
+      setAttempt({ status: 'failed', statusText: e.message })
+    );
+}
+```
+
+### useAsync
+
+This is a newer custom hook that is used to manage async state in Teleport. It requires less boilerplate than
+`useAttempt`, as it handles running async operations and handling errors, but still requires a manual trigger in an
+effect,
+which causes data to not fetch until at least the component has rendered once.
+
+Whilst this works fine, it lacks an extended API that TanStack Query can provide.
+
+#### Example query
+
+```tsx
+const [fetchQrCodeAttempt, fetchQrCode] = useAsync((privilegeToken: string) =>
+  auth.createMfaRegistrationChallenge(privilegeToken, 'totp')
+);
+
+useEffect(() => {
+  fetchQrCode(privilegeToken);
+}, []);
+```
+
+#### Example mutation
+
+```tsx
+const [deleteRequestAttempt, runDeleteRequest] = useAsync(async () => {
+  await ctx.workflowService.deleteAccessRequest(props.requestId);
+  historyService.replace(cfg.getAccessRequestRoute());
+});
+```
+
+## TanStack Query
+
+TanStack Query offers many powerful abilities, including:
+
+- Automatic retries and manual re-fetching methods
+  - Automatic retries can be configured to retry failed requests a certain number of times, with exponential backoff, something not easily achievable
+    with `useAttempt` or `useAsync`
+- Caching and background re-fetching
+  - There are no caching abilities on the frontend as it stands, so avoiding unnecessary network requests is a big win
+- Infinite queries
+  - This can allow us to create infinite scrolling tables, which is a better UX for larger customers
+- Paginated queries
+  - At the moment we have a large wrapper utility to add paginated query support for our tables. TanStack Query provides a built-in
```

Comment:
Will it integrate with our existing gRPC list APIs or will we have to rewrite them?

Every time we touch pagination code we introduce bugs, so this point only seems like a positive to me if it doesn't require us to change all of our APIs.

Reply:
I don't see any changes for the backend needed, it should work fine with the way we currently paginate things (data, next_cursor, but any shape works really - it's very flexible)